dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,517 @@
+        """
+        req_batch_id = pipe_run_ref.batch_id
+
+        # In case we haven't scheduled the iteration yet (i.e. deterministic config), do it now
+        if not pipe_run_ref.is_scheduled:
+            self._schedule_batch(pipe_run_ref)
+            pipe_run_ref.is_scheduled = True
+
+        # Wait for the requested output to be ready
+        req_outputs = None
+        # If the data was already read, just return it (and clear the cache entry)
+        if req_batch_id in self.cache_outputs:
+            req_outputs = self.cache_outputs[req_batch_id]
+            del self.cache_outputs[req_batch_id]
+
+        else:
+            curr_batch_id = None
+            # If not the data we are looking for, store it and keep processing until we find it
+            while req_batch_id != curr_batch_id:
+                torch.cuda.nvtx.range_push("dali_output_q.get")
+                curr_batch_id, curr_processed_outputs, err = self.dali_output_q.get()
+                torch.cuda.nvtx.range_pop()
+
+                if err is not None:
+                    raise err
+
+                if curr_batch_id == req_batch_id:
+                    req_outputs = curr_processed_outputs
+                else:
+                    self.cache_outputs[curr_batch_id] = curr_processed_outputs
+        # Unpack single element tuples
+        if isinstance(req_outputs, tuple) and len(req_outputs) == 1:
+            req_outputs = req_outputs[0]
+        return req_outputs
+
+    def produce_data(self, obj):
+        """
+        A generic function to recursively visits all elements in a nested structure and replace
+        instances of DALIPipelineRunRef with the actual data provided by the DALI server
+
+        Args:
+            obj: The object to map (can be an instance of any class).
+
+        Returns:
+            A new object where any instance of DALIPipelineRunRef has been replaced with actual
+            data.
+        """
+        # If it is an instance of DALIPipelineRunRef, replace it with data
+        if isinstance(obj, DALIPipelineRunRef):
+            return self._get_outputs(obj)
+        # If it is a custom class, recursively call produce data on its members
+        elif hasattr(obj, "__dict__"):
+            for attr_name, attr_value in obj.__dict__.items():
+                setattr(obj, attr_name, self.produce_data(attr_value))
+            return obj
+
+        # If it's a list, recursively apply the function to each element
+        elif isinstance(obj, list):
+            return [self.produce_data(item) for item in obj]
+
+        # If it's a tuple, recursively apply the function to each element (and preserve tuple type)
+        elif isinstance(obj, tuple):
+            return tuple(self.produce_data(item) for item in obj)
+
+        # If it's a dictionary, apply the function to both keys and values
+        elif isinstance(obj, dict):
+            return {key: self.produce_data(value) for key, value in obj.items()}
```

Comment:
Can we do anything as wild as force Python to return mapping from `DALIPipelineRunRef` to `DALIPipelineRunRef` somehow? How would one construct such proxy?